pkg/security/podsecuritypolicy/seccomp/strategy.go

```diff
@@ -110,22 +140,55 @@ func (s *strategy) ValidatePod(pod *api.Pod) field.ErrorList {
 // the range of the strategy.
 func (s *strategy) ValidateContainer(pod *api.Pod, container *api.Container) field.ErrorList {
 	allErrs := field.ErrorList{}
-	fieldPath := field.NewPath("pod", "metadata", "annotations").Key(api.SeccompContainerAnnotationKeyPrefix + container.Name)
-	containerProfile := profileForContainer(pod, container)
 
-	if !s.allowAnyProfile && len(s.allowedProfiles) == 0 && containerProfile != "" {
-		allErrs = append(allErrs, field.Forbidden(fieldPath, "seccomp may not be set"))
-		return allErrs
+	// Receive necessary values and do pre-checks
+	annotationFieldPath := field.NewPath("pod", "metadata", "annotations").Key(api.SeccompContainerAnnotationKeyPrefix + container.Name)
+	profileAnnotation, hasAnnotation := profileAnnotationForContainer(pod, container)
+	hasField := (container.SecurityContext != nil && container.SecurityContext.SeccompProfile != nil)
+
+	if !s.allowAnyProfile {
+		if len(s.allowedProfiles) == 0 && (hasAnnotation || hasField) {
+			allErrs = append(allErrs, field.Forbidden(annotationFieldPath, "seccomp may not be set"))
+			return allErrs
+		}
+
+		if len(s.allowedProfiles) != 0 && !hasAnnotation && !hasField {
+			allErrs = append(allErrs, field.Forbidden(annotationFieldPath, s.notAllowedProfileMsg("")))
+			return allErrs
+		}
 	}
 
-	if !s.profileAllowed(containerProfile) {
-		msg := fmt.Sprintf("%s is not an allowed seccomp profile. Valid values are %v", containerProfile, s.allowedProfilesString)
-		allErrs = append(allErrs, field.Forbidden(fieldPath, msg))
+	// Annotation based validation
+	if hasAnnotation && !s.profileAllowed(profileAnnotation) {
+		allErrs = append(allErrs, field.Forbidden(annotationFieldPath, s.notAllowedProfileMsg(profileAnnotation)))
+	}
+
+	// Field based validation
+	securityContextFldPath := field.NewPath("pod", "spec", "containers[*]", "securityContext", "seccompProfile", "type")
```

Comment:
This may not always be correct, we need to account for `initContainers` too. Ideally this would have a similar interface as `VisitContainersWithPath`, receiving the field path as parameter.

Reply:
We do not have something like this at hand, right? Or do you see any utility method which can be used?

Reply:
This discussion is now obsolete because of the changed implementation approach.